File: 1539-kth-missing-positive-number/1539-kth-missing-positive-number.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number}
 */
var findKthPositive = function (arr, k) {
    const n = arr.length;
    // writing condition for the missing number to be prresent in the right side

    if (arr[n - 1] - n < k) {
        return k+n
    }

    // what if the missing number lies in the left side

    if (arr[0] > k) {
        return k;
    }

    // now the only condition what if it lies  in the middle of the array

    let low = 0;
    let high = n - 1;
    let val;
    while (low <= high) {
        let mid = Math.floor((low + high) / 2);

        if (arr[mid] - (mid + 1) < k) {
            low = mid + 1;
        }
        else {
            val = mid;
            high = mid - 1;
        }
       
    }
    console.log({ val })

    return arr[val] - (arr[val] - (val + 1) - k + 1);

};

File: 1539-kth-missing-positive-number/1539-kth-missing-positive-number.cpp
class Solution {
public:
    int findKthPositive(vector<int>& arr, int k) {
        int n = arr.size();
        int index =0;
        int value =1;
        int oldk=k;
        while(index <n && k>0){
            if (arr[index]== value){
                index++;
                value++;
            }
            else{
                value++;
                k--;
            }
        }
       
        if(k>0){
            int diff = arr[n-1]- n;
            return arr[n-1] - diff + oldk;
        }
        return value-1;
    }
};